### ISSUE TITLE
Validate config before running test

### ISSUE THREAD

**Issue by marseel** [kubernetes/perf-tests#1636]
**What would you like to be added**:
Validate if config is correct before running test.
**Why is this needed**:
Currently, step can be either measurement step or phase step (modules are being added in #1634)
https://github.com/kubernetes/perf-tests/blob/18e90fc65c6b95c0bff96938458a77d74e19b27a/clusterloader2/api/types.go#L60
Unfortunately, this behaviour is not well implemented:
https://github.com/kubernetes/perf-tests/blob/14afe3828ba8d9eaa98aa8a9f5ddd0cdc5f8eebf/clusterloader2/api/extensions.go#L54
https://github.com/kubernetes/perf-tests/blob/14afe3828ba8d9eaa98aa8a9f5ddd0cdc5f8eebf/clusterloader2/pkg/test/simple_test_executor.go#L143
Current implementation does not validate if config is correct before running test.
From user experience perspective it would be much better to return error right after test starts. 
Also, If I'm not wrong, if user specifies both measurements and phases, only measurements will be executed.

**Comment by lyzs90** [comment]
@marseel could I help out with this one?

**Comment by marseel** [comment]
@lyzs90  Yes sure, I would appreciate it :) 
If you will have some PR I can help with reviewing.

**Comment by wojtek-t** [comment]
@lyzs90 - this long-standing PR is relevant to this one:
https://github.com/kubernetes/perf-tests/pull/142
It shows how should probably approach this problem - feel free to pick it up and continue working on that.

**Comment by lyzs90** [comment]
Thanks @wojtek-t I'll check it out

/assign

**Comment by lyzs90** [comment]
@wojtek-t I had a look at https://github.com/kubernetes/perf-tests/pull/142 and had some thoughts:
- Currently Modules are being recursively compiled into steps at [runtime](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/simple_test_executor.go#L125). Correct me if I am wrong but the mapping passed to the Module seems static and there is no use case where runtime variables are passed in (unlike Objects). Hence to support validation of Modules, may I suggest to move loading of modules before test execution - after https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L66, so that validation of the entire Config (less Objects) can be validated ahead of time
- I understand that the current style of validation used to keep consistent with k8s, and I'm happy to stick with it. Just wanted to find out what are your thoughts on using a declarative approach like json schema? https://github.com/xeipuuv/gojsonschema
  - Checks like [this](https://github.com/kubernetes/perf-tests/blob/ab39cc7036cdb082588ecf5522f8f3f897eb4ca3/clusterloader2/api/validation/validation.go#L99) and [this](https://github.com/kubernetes/perf-tests/blob/ab39cc7036cdb082588ecf5522f8f3f897eb4ca3/clusterloader2/api/validation/validation.go#L43) could be simplified with https://json-schema.org/understanding-json-schema/reference/combining.html#combining-schemas
  - Basically cuts down on boilerplate, but we may still have to implement some custom validation for things like IsDNS1123Subdomain, file exists at objectTemplatePath, referenced tuning set has been declared etc.

**Comment by wojtek-t** [comment]
> Currently Modules are being recursively compiled into steps at runtime. Correct me if I am wrong but the mapping passed to the Module seems static and there is no use case where runtime variables are passed in (unlike Objects). Hence to support validation of Modules, may I suggest to move loading of modules before test execution - after https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L66, so that validation of the entire Config (less Objects) can be validated ahead of time

@mm4tt - for thoughts

> I understand that the current style of validation used to keep consistent with k8s, and I'm happy to stick with it. Just wanted to find out what are your thoughts on using a declarative approach like json schema? https://github.com/xeipuuv/gojsonschema

The consistency with k8s has a benefit that you don't have to learn new stuff when you're operating in a single k8s ecosystem. I agree that this can reduce the boilerplate, but (a) this boilerplate is well-separated (b) it's fairly trivial, so I don't think it's actually that huge advantage. So I would rather stick to that for now.

**Comment by lyzs90** [comment]
> The consistency with k8s has a benefit that you don't have to learn new stuff when you're operating in a single k8s ecosystem. I agree that this can reduce the boilerplate, but (a) this boilerplate is well-separated (b) it's fairly trivial, so I don't think it's actually that huge advantage. So I would rather stick to that for now.

Noted on this 👍

**Comment by marseel** [comment]
> Currently Modules are being recursively compiled into steps at runtime. Correct me if I am wrong but the mapping passed to the Module seems static and there is no use case where runtime variables are passed in (unlike Objects). Hence to support validation of Modules, may I suggest to move loading of modules before test execution - after https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L66, so that validation of the entire Config (less Objects) can be validated ahead of time
I think it makes sense, but let's double check with @mm4tt

**Comment by mm4tt** [comment]
It makes a lot of sense :) Just a nit, you should move it after [this](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L76) to make sure that the config is valid after a custom modifications have been applied. There is also testConfig.Validate [here](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L89) that we could probably reuse or remove - it doesn't make sense to have two "validate" components. 

In general, big +1 for extracting things out of ExecuteTest method. In my opinion, this method is too big. It does too many things and doesn't adhere to the single principle rule - it should just execute the test that is prepared and validated.

**Comment by lyzs90** [comment]
@mm4tt gotcha, I'll first submit a PR to extract module compilation out of ExecuteTest

**Comment by lyzs90** [comment]
@mm4tt Actually since there might be multiple tests (via test suite / config paths), would it be worth moving the compilation and validation logic even higher up eg. before https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/cmd/clusterloader.go#L317 and https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/cmd/clusterloader.go#L323. 

That way we can fail faster if any config is invalid. This would mean [RunTest](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/test/test.go#L41:6) would only be left with basic checks, namespace deletion and test execution

**Comment by mm4tt** [comment]
Sounds good. Let me know if there is anything I can help you with. Thanks!
